tests/samplers_tests/test_nsgaii.py

```diff
@@ -157,6 +159,151 @@ def constraints_func(trial: FrozenTrial) -> Sequence[float]:
+        for c2 in constraints_1d_infeasible
+    ]
+
+    # In the following code, we test that the feasible trials always dominate
+    # the infeasible trials.
+    for constraints1 in constraints_list1:
+        for constraints2 in constraints_list2:
+            t1 = _create_frozen_trial(0, [0], constraints1)
+            t2 = _create_frozen_trial(1, [1], constraints2)
+            assert _constrained_dominates(t1, t2, directions)
+            assert not _constrained_dominates(t2, t1, directions)
+
+            t1 = _create_frozen_trial(0, [1], constraints1)
+            t2 = _create_frozen_trial(1, [0], constraints2)
+            assert _constrained_dominates(t1, t2, directions)
+            assert not _constrained_dominates(t2, t1, directions)
+
+
+@pytest.mark.parametrize("direction", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
+def test_constrained_dominates_infeasible_vs_infeasible(direction: StudyDirection) -> None:
+    inf = float("inf")
+    directions = [direction]
+
+    # The following table illustrates the violations of some constraint values.
+    # When both trials are infeasible, the trial with smaller violation dominates
+    # the one with larger violation.
+    #
+    #                       c2
+    #      ╔═════╤═════╤═════╤═════╤═════╤═════╗
+    #      ║     │ -1  │  0  │  1  │  2  │  ∞  ║
+    #      ╟─────┼─────┼─────┼─────┼─────┼─────╢
+    #      ║ -1  │           │  1  │  2  │  ∞  ║
+    #      ╟─────┼─feasible ─┼─────┼─────┼─────╢
+    #      ║  0  │           │  1  │  2  │  ∞  ║
+    # c1   ╟─────┼─────┼─────┼─────┼─────┼─────╢
+    #      ║  1  │  1  │  1  │  2  │  3  │  ∞  ║
+    #      ╟─────┼─────┼─────┼─────┼─────┼─────╢
+    #      ║  2  │  2  │  2  │  3  │  4  │  ∞  ║
+    #      ╟─────┼─────┼─────┼─────┼─────┼─────╢
+    #      ║  ∞  │  ∞  │  ∞  │  ∞  │  ∞  │  ∞  ║
+    #      ╚═════╧═════╧═════╧═════╧═════╧═════╝
+    #
+
+    # Check all pairs of these constraints.
+    constraints_infeasible_sorted: List[List[List[float]]]
+    constraints_infeasible_sorted = [
+        # These constraints have violation 1.
+        [[1, -inf], [1, -1], [1, 0], [0, 1], [-1, 1], [-inf, 1]],
+        # These constraints have violation 2.
+        [[2, -inf], [2, -1], [2, 0], [1, 1], [0, 2], [-1, 2], [-inf, 2]],
+        # These constraints have violation 3.
+        [[3, -inf], [3, -1], [3, 0], [2, 1], [1, 2], [0, 3], [-1, 3], [-inf, 3]],
+        # These constraints have violation inf.
+        [
+            [-inf, inf],
+            [-1, inf],
+            [0, inf],
+            [1, inf],
+            [inf, inf],
+            [inf, 1],
+            [inf, 0],
+            [inf, -1],
+            [inf, -inf],
+        ],
+    ]
+
+    # Check that constraints with smaller violations dominate constraints with larger violation.
+    for i in range(len(constraints_infeasible_sorted)):
+        for j in range(i + 1, len(constraints_infeasible_sorted)):
```

Comment:
[FYI] Now I'm thinking that how we can refactor these loop conditions...

Reply:
For example, we may check these cases like:

```suggestion
    constraints_infeasible_violation1 = [[1, -inf], [1, -1], [1, 0], [0, 1], [-1, 1], [-inf, 1]]
    constraints_infeasible_violation2 = [[1, -inf], [1, -1], [1, 0], [0, 1], [-1, 1], [-inf, 1]]
    for constraints1 in constraints_infeasible_violation1:
        for constraints2 in constraints_infeasible_violation2:
            t1 = ...
            ...

    constraints_infeasible_violation1 = [[1, -inf], [1, -1], [1, 0], [0, 1], [-1, 1], [-inf, 1]]
    constraints_infeasible_violation_inf = [[1, -inf], [1, -1], [1, 0], [0, 1], [-1, 1], [-inf, 1]]
    for constraints1 in constraints_infeasible_violation1:
        for constraints2 in constraints_infeasible_violation_inf:
            t1 = ...
            ...            
```

Reply:
My suggested change is a bit controversial in terms of test coverages. From our offline discussion with @contramundum53, we decided not to change these code. Since the main goal of this PR is to add tests, I think it is reasonable to keep any further refactoring out of scope.